### PR TITLE
Implemented Tags Suggestion Feature

### LINKED ISSUE #5: Tags suggestion
Tags in Song View Controller should have autocomplete or suggestion feature based on the tags already available. The goal of this enhancement is to prevent typos in writing pre-existing tags

### PR DESCRIPTION
**Pull request by prabogey**
From [github issue](https://github.com/youstanzr/YouTag/issues/5)

Added SearchTextField library to provide textField with dropdown
capabilities.

### CODE REVIEW

File: YouTag/YYTTagCell.swift [hunk 1]
```diff
@@ -73,8 +79,24 @@ class YYTTagCell: UICollectionViewCell {
 		textField.leadingAnchor.constraint(equalTo: self.contentView.leadingAnchor, constant: 2.5).isActive = true
 		textField.bottomAnchor.constraint(equalTo: self.contentView.bottomAnchor, constant: -2.5).isActive = true
 		textField.topAnchor.constraint(equalTo: self.contentView.topAnchor, constant: 6).isActive = true  // Because the font is shifted upward
+        textField.addTarget(self, action: #selector(YYTTagCell.textChanged), for: .editingChanged) // adding text change selector
 	}
-	
+    
+    // This function will fire whenever the user is typing out a new tag to add to a song
+    @objc func textChanged(txtField: UITextField) {
+        let tagList = LibraryManager.getAll(.tags)
```
Comment (youstanzr) [inline YouTag/YYTTagCell.swift]: This would make artist tagView suggest tags instead of artists. Actually any tagView will have a suggestion of artist no matter data it should be suggesting.

File: YouTag/YYTTagCell.swift [hunk 2]
```diff
@@ -45,6 +49,8 @@ class YYTTagCell: UICollectionViewCell {
 		super.init(frame: frame)
 		self.layer.cornerRadius = self.frame.height / 2.0
 		self.layer.borderWidth = 2.0
+        self.clipsToBounds = false
+        self.contentView.clipsToBounds = false
```
Comment (youstanzr) [inline YouTag/YYTTagCell.swift]: Why is this needed?

File: YouTag/YYTTagView.swift [hunk 3]
```diff
@@ -201,6 +201,7 @@ class YYTTagView: UICollectionView, UICollectionViewDataSource, UICollectionView
 			tagsList.add(textField.text!.capitalized)
 			self.reloadData()
 		}
+//        UIApplication.shared.keyWindow!.sendSubviewToBack(textField.superview?.superview as! YYTTagCell)
```
Comment (youstanzr) [inline YouTag/YYTTagView.swift]: I thought this was necessary for my code but I don't see any impact after you removed it. Well, I guess its fine to remove it

File: YouTag/YYTTagCell.swift [hunk 4]
```diff
@@ -73,8 +79,24 @@ class YYTTagCell: UICollectionViewCell {
...
+    @objc func textChanged(txtField: UITextField) {
+        let tagList = LibraryManager.getAll(.tags)
+        for i in 0 ..< tagList.count {
+            print("Tag: \(tagList[i])")
+        }
+        if (txtField.text!.length == 0) {
+            suggestionTableView.isHidden = true
+        } else {
+            suggestionTableView.isHidden = false
+        }
```
Comment (youstanzr) [inline YouTag/YYTTagCell.swift]: `suggestionTableView.isHidden = (txtField.text!.length == 0) ? true : false` or we can just make it equal to the condition directly without the ternary operator.

File: YouTag/YYTTagCell.swift [hunk 5]
```diff
@@ -73,8 +79,24 @@ class YYTTagCell: UICollectionViewCell {
...
+        textField.addTarget(self, action: #selector(YYTTagCell.textChanged), for: .editingChanged) // adding text change selector
 	}
-	
+    
+    // This function will fire whenever the user is typing out a new tag to add to a song
+    @objc func textChanged(txtField: UITextField) {
+        let tagList = LibraryManager.getAll(.tags)
+        for i in 0 ..< tagList.count {
+            print("Tag: \(tagList[i])")
+        }
```
Comment (youstanzr) [inline YouTag/YYTTagCell.swift]: This can be removed

File: YouTag/SongDetailViewController.swift [hunk 6]
```diff
@@ -8,6 +8,7 @@
 
 import UIKit
 
+// This is the view that shows up whenever you add a new song from the web and want to edit its details before adding to your library
```
Comment (youstanzr) [inline YouTag/SongDetailViewController.swift:11]: 👍

File: YouTag/YYTTagCell.swift [hunk 7]
```diff
@@ -17,15 +17,19 @@ class YYTTagCell: UICollectionViewCell {
...
 		txtfld.tintColor = .white
 		txtfld.backgroundColor = UIColor.clear
 		txtfld.autocorrectionType = .no
 		txtfld.returnKeyType = .done
 		return txtfld
 	}()
+    let suggestionTableView: UITableView = {
+        let tableView = UITableView()
+        return tableView
+    }()
```
Comment (youstanzr) [inline YouTag/YYTTagCell.swift]: I think the SearchTextField library has its own tableView and we don't need to create one, do we?

**Review by youstanzr — COMMENTED** [review]
First, I'd like to thank you for the contribution.

I think we can improve this by making tag suggestion optional for an addEnabled tagview and if the suggestion option was set to yes then we supply the suggestion data source. This way would also help us remove the heavy computation to tag each time the text changes.

If you'd like to step it up a notch, then you can create those suggestion appear above the keyboard in a horizontally scrollable bar style just like Apple's keyboard suggestions. This would be really nice UI and provide better UX.

File: YouTag/YYTTagCell.swift (hunk 7 above)
Comment (prabogey) [reply YouTag/YYTTagCell.swift]: Yes you are right I'll change this.

**Comment by prabogey** [comment]
How do you suggest we ask for them to turn on suggestions or not?

**Comment by youstanzr** [comment]
It should not be a user input kinda choice. I meant add a new class variable for YTTagView that you can enable or disable suggestions.
We can make the variable private and have a special setter function to turn it on and that function would have a parameter of type array which is going to act like the data source for the suggestions. Also if we do it this way, then we have to create another function that will turn suggestions off.
Another way to do this would be to make the variable public and have another optional variable that will hold the suggestion array and handle it from there.

**Comment by prabogey** [comment]
Im confused as to why this is necessary, suggestions would only be turned on if the user is typing right? Therefore, if the user is ever typing then they will need the suggestions.

**Comment by youstanzr** [comment]
In my opinion, suggestions should not be forced when you create a tagView with isAddEnabled property. I'd rather have it optional for the object so we have better reusable code. Same goes for the suggestion array, I don't want to have dependency of LibraryManager inside of the tag cell which is going to kill reusability. We can already see how it causes trouble to have the suggestion data source hardcoded in the tag cell by looking at the artist tag view and how its forced to use the tag array as its source.

**Comment by prabogey** [comment]
Understood, ill go ahead and add a class variable for that when a class calls YTTagView they can set the suggestionsEnabled as well as if the tag view is for artists or for genres.

**Comment by youstanzr** [comment]
Just make the suggestion data source array as a variable too so we can use it for everything instead of locking it to artists or tags. Then when we have the tags tagView, we pass the `getAll(.tags)`and when we have artist we pass `getAll(.artists)`. If we later introduced more tagView properties to song, then we just pass the new suggestion array and that clears everything up for us.

**Comment by prabogey** [comment]
It turns out YTTagView takes in a tag list already so I will just re-use that and pass it down to the CellView.

**Comment by youstanzr** [comment]
YYTTagView tag list is used as the data source for tags and it should not be used as the data source for suggestions.

**Comment by prabogey** [comment]
Aren't the suggestions the same as the list of tags? The library handles showing suggestions that are more closely related to what has been typed already. They handle the algorithm that compares the similarity of what has been typed to the suggestions, all we have to do is provide them with a list of all possible suggestions when the view is created.

**Comment by youstanzr** [comment]
YYTagView displays tags based on tag list. So, if you add a tag, it's added to tags list. A tag list can be empty for a particular song but suggestions would still appear from the other tags available from the other song. If you assume that tag list is the suggestion list, then it will never suggest anything that is not part of the added list, which actually it should not suggest.

Now when I am thinking about it, actually you should always remove the tag list elements from the suggestion list as you don't want to display a tag that was already in added.

**Comment by youstanzr** [comment]
Also, don't forget to add the the library in the attributions in the Readme file

**Comment by prabogey** [comment]
I see, I missed that part in code. A lot of callers to YTTagView actually get the entire list of tags from the LibraryManager but since not all do I understand why we can't use that. 

As per your other concern, I thought the intended purpose is that we wanted to show tags from all songs on the library so that the user does not mistype new tags it wants to add to a song.

**Comment by youstanzr** [comment]
Exactly, but why do you want to suggest the user to add a tag that is already added. Thats just suggesting the user to add a duplicate item.

**Comment by prabogey** [comment]
Yes that part I will write a function to remove all current tags from the list of all tags.

**Comment by youstanzr** [comment]
Great 👍 Just make sure to call the function in the didSet of the suggestion data source array

**Comment by prabogey** [comment]
The additions should be there now, the autocomplete works as intended and does not supply duplicates.

File: YouTag/FilterPickerView.swift [hunk 8]
```diff
@@ -152,7 +152,7 @@ class FilterPickerView: UIView {
 		pickerView.topAnchor.constraint(equalTo: filterSegment.bottomAnchor).isActive = true
 		pickerView.bottomAnchor.constraint(equalTo: addButton.topAnchor).isActive = true
 		
-		tagView = YYTTagView(frame: .zero, tagsList: LibraryManager.getAll(.tags), isAddEnabled: false, isMultiSelection: true, isDeleteEnabled: false)
+		tagView = YYTTagView(frame: .zero, tagsList: LibraryManager.getAll(.tags), isAddEnabled: false, isMultiSelection: true, isDeleteEnabled: false, areSuggestionsEnabled: true, isForArtists: false)
```
Comment (youstanzr) [inline YouTag/FilterPickerView.swift]: Change isForArtist to be suggestionDataSource which should be optional

File: YouTag/SongDetailViewController.swift [hunk 9]
```diff
@@ -136,7 +136,7 @@ class SongDetailViewController: UIViewController, UITextFieldDelegate, UITextVie
 		titleTextField.widthAnchor.constraint(equalTo: self.view.widthAnchor, multiplier: 0.8).isActive = true
 
 		let artistsTags = NSMutableArray(array: songDict["artists"] as? NSArray ?? NSArray())
-		artistsTagsView = YYTTagView(frame: .zero, tagsList: artistsTags, isAddEnabled: true, isMultiSelection: false, isDeleteEnabled: true)
+        artistsTagsView = YYTTagView(frame: .zero, tagsList: artistsTags, isAddEnabled: true, isMultiSelection: false, isDeleteEnabled: true, areSuggestionsEnabled: true, isForArtists: true)
```
Comment (youstanzr) [inline YouTag/SongDetailViewController.swift]: Make sure indentation is proper

File: YouTag/YYTFilterTagView.swift [hunk 10]
```diff
@@ -10,8 +10,8 @@ import UIKit
 
 class YYTFilterTagView: YYTTagView {
 	
-	override init(frame: CGRect, tagsList: NSMutableArray, isAddEnabled: Bool, isMultiSelection: Bool, isDeleteEnabled: Bool) {
-		super.init(frame: frame, tagsList: tagsList, isAddEnabled: isAddEnabled, isMultiSelection: isMultiSelection, isDeleteEnabled: isDeleteEnabled)
+    override init(frame: CGRect, tagsList: NSMutableArray, isAddEnabled: Bool, isMultiSelection: Bool, isDeleteEnabled: Bool, areSuggestionsEnabled: Bool, isForArtists: Bool) {
+        super.init(frame: frame, tagsList: tagsList, isAddEnabled: isAddEnabled, isMultiSelection: isMultiSelection, isDeleteEnabled: isDeleteEnabled, areSuggestionsEnabled: areSuggestionsEnabled, isForArtists: isForArtists)
```
Comment (youstanzr) [inline YouTag/YYTFilterTagView.swift]: Should be like this:
```
    override init(frame: CGRect, tagsList: NSMutableArray, isAddEnabled: Bool, isMultiSelection: Bool, isDeleteEnabled: Bool) {
         super.init(frame: frame, tagsList: tagsList, isAddEnabled: isAddEnabled, isMultiSelection: isMultiSelection, isDeleteEnabled: isDeleteEnabled, areSuggestionsEnabled: false, suggestionArray: nil)

```

Comment (youstanzr) [reply YouTag/YYTFilterTagView.swift]: I just thought about an improvement to this init. I'll do it now and pull it so you can use that

File: YouTag/YYTTagCell.swift [hunk 11]
```diff
@@ -44,13 +40,21 @@ class YYTTagCell: UICollectionViewCell {
...
 	fileprivate var titleLabelDefaultLeadingAnchor: NSLayoutConstraint?
 	fileprivate var titleLabelWithImageLeadingAnchor: NSLayoutConstraint?
+    fileprivate var tagList: [String]?
+    var areSuggestionsEnabled: Bool? {
+        didSet {
+            if ((tagList != nil) && (areSuggestionsEnabled!)) {
+                textField.filterStrings(tagList ?? [""])
+            } else {
+                textField.filterStrings([])
+            }
```
Comment (youstanzr) [inline YouTag/YYTTagCell.swift]: Make sure indentation is proper

File: YouTag/YYTTagView.swift [hunk 12]
```diff
@@ -17,17 +17,24 @@ class YYTTagView: UICollectionView, UICollectionViewDataSource, UICollectionView
 	weak var yytdelegate: YYTTagViewDelegate?
 	var addTagPlaceHolder: String!
 	var isAddEnabled: Bool!
+    var areSuggestionsEnabled: Bool!
+    var isForArtists: Bool!
```
Comment (youstanzr) [inline YouTag/YYTTagView.swift]: This should be removed

File: YouTag/YYTTagView.swift [hunk 13]
```diff
@@ -17,17 +17,24 @@ class YYTTagView: UICollectionView, UICollectionViewDataSource, UICollectionView
...
+        didSet {
+            self.suggestionsList = self.suggestionsList?.filter{ !self.tagsList.contains($0) }
+        }
+    }
 	var selectedTagList = NSMutableArray()
 	var isEditing: Bool = false
 	
 	
-	init(frame: CGRect, tagsList: NSMutableArray, isAddEnabled: Bool, isMultiSelection: Bool, isDeleteEnabled: Bool) {
+    init(frame: CGRect, tagsList: NSMutableArray, isAddEnabled: Bool, isMultiSelection: Bool, isDeleteEnabled: Bool, areSuggestionsEnabled: Bool, isForArtists: Bool) {
```
Comment (youstanzr) [inline YouTag/YYTTagView.swift]: indentation fix

File: YouTag/YYTTagCell.swift [hunk 14]
```diff
@@ -79,24 +83,8 @@ class YYTTagCell: UICollectionViewCell {
...
-    // This function will fire whenever the user is typing out a new tag to add to a song
-    @objc func textChanged(txtField: UITextField) {
-        let tagList = LibraryManager.getAll(.tags)
-        for i in 0 ..< tagList.count {
-            print("Tag: \(tagList[i])")
-        }
-        if (txtField.text!.length == 0) {
-            suggestionTableView.isHidden = true
-        } else {
-            suggestionTableView.isHidden = false
```
Comment (youstanzr) [inline YouTag/YYTTagCell.swift]: I think removing this will break the suggestion code as you still have to use filterStrings function each time the array changes

File: YouTag/YYTTagCell.swift (hunk 11 above)
Comment (youstanzr) [reply YouTag/YYTTagCell.swift]: Why do we have a tag list for each cell. That not needed. Also, `areSuggestionsEnabled` didSet should not filter the string but it should affect whether the suggestion tableview would appear or not. Filtering should be at textChange func for the textfield.

File: YouTag/YYTTagCell.swift [hunk 15]
```diff
@@ -121,4 +109,7 @@ class YYTTagCell: UICollectionViewCell {
 		}
 	}
 	
+    func setTagList(tagList: [String]) {
+        self.tagList = tagList
```
Comment (youstanzr) [inline YouTag/YYTTagCell.swift]: Not needed

File: YouTag/YYTTagView.swift [hunk 16]
```diff
@@ -17,17 +17,24 @@ class YYTTagView: UICollectionView, UICollectionViewDataSource, UICollectionView
 	weak var yytdelegate: YYTTagViewDelegate?
 	var addTagPlaceHolder: String!
 	var isAddEnabled: Bool!
+    var areSuggestionsEnabled: Bool!
```
Comment (youstanzr) [inline YouTag/YYTTagView.swift]: Fix indentation

File: YouTag/YYTTagView.swift [hunk 17]
```diff
@@ -17,17 +17,24 @@ class YYTTagView: UICollectionView, UICollectionViewDataSource, UICollectionView
...
 	var tagsList: NSMutableArray! {
 		didSet {
 			selectedTagList.removeAllObjects()
 		}
 	}
+    var suggestionsList: [String]? {
+        didSet {
+            self.suggestionsList = self.suggestionsList?.filter{ !self.tagsList.contains($0) }
+        }
+    }
```
Comment (youstanzr) [inline YouTag/YYTTagView.swift]: Fix indentation

File: YouTag/YYTTagView.swift [hunk 18]
```diff
@@ -44,6 +51,8 @@ class YYTTagView: UICollectionView, UICollectionViewDataSource, UICollectionView
 		self.isAddEnabled = isAddEnabled
 		self.isDeleteEnabled = isDeleteEnabled
 		self.tagsList = tagsList
+        self.areSuggestionsEnabled = areSuggestionsEnabled
+        self.isForArtists = isForArtists
```
Comment (youstanzr) [inline YouTag/YYTTagView.swift]: Fix indentation

File: YouTag/YYTTagCell.swift (hunk 11 above)
Comment (prabogey) [reply YouTag/YYTTagCell.swift]: This is not doing any filtering, it is simply setting the list of all possible drop downs in the SearchTextField class - the method name is a bit misleading.

File: YouTag/YYTTagView.swift [hunk 19]
```diff
@@ -166,6 +175,13 @@ class YYTTagView: UICollectionView, UICollectionViewDataSource, UICollectionView
 			print("Add Tag Button tapped")
 			isEditing = true
 			cell.switchMode(enableEditing: true)
+            if self.isForArtists {
+                self.suggestionsList = LibraryManager.getAll(.artists) as? [String]
+            } else {
+                self.suggestionsList = LibraryManager.getAll(.tags) as? [String]
+            }
+            cell.setTagList(tagList: self.suggestionsList!)
```
Comment (youstanzr) [inline YouTag/YYTTagView.swift]: This is an optional value by setting and should not be force unwrapped

File: YouTag/YYTTagCell.swift (hunk 11 above)
Comment (youstanzr) [reply YouTag/YYTTagCell.swift]: Thats really a misleading name but alright for `areSuggestionsEnabled` all good. Still the tagList should not be a member of the cell

File: YouTag/YYTTagView.swift [hunk 20]
```diff
@@ -23,11 +23,18 @@ class YYTTagView: UICollectionView, UICollectionViewDataSource, UICollectionView
 			selectedTagList.removeAllObjects()
 		}
 	}
+    var suggestionsList: [String]? {
+        didSet {
+            if let _ = suggestionsList {
+                self.suggestionsList = self.suggestionsList?.filter{ !self.tagsList.contains($0) }
+            }
```
Comment (youstanzr) [inline YouTag/YYTTagView.swift]: You should also add this in tagList didSet so that suggestionList will get updated when tagList gets changed

File: YouTag/YYTTagView.swift [hunk 21]
```diff
@@ -23,11 +23,18 @@ class YYTTagView: UICollectionView, UICollectionViewDataSource, UICollectionView
...
+        }
+    }
 	var selectedTagList = NSMutableArray()
 	var isEditing: Bool = false
 	
 	
-	init(frame: CGRect, tagsList: NSMutableArray, isAddEnabled: Bool, isMultiSelection: Bool, isDeleteEnabled: Bool) {
+    init(frame: CGRect, tagsList: NSMutableArray, isAddEnabled: Bool, isMultiSelection: Bool, isDeleteEnabled: Bool, suggestionDataSource: [String]?) {
```
Comment (youstanzr) [inline YouTag/YYTTagView.swift]: I noticed that through out the code you're using spaces for the indentation instead of tabs. Could you please change that to keep the overall code style consistent. This line is just an example of this space vs tabs indentation

File: YouTag/YYTTagCell.swift [hunk 22]
```diff
@@ -40,6 +40,14 @@ class YYTTagCell: UICollectionViewCell {
 	
 	fileprivate var titleLabelDefaultLeadingAnchor: NSLayoutConstraint?
 	fileprivate var titleLabelWithImageLeadingAnchor: NSLayoutConstraint?
+    fileprivate var tagList: [String]?
+    var areSuggestionsEnabled: Bool? {
```
Comment (youstanzr) [inline YouTag/YYTTagCell.swift]: Those variables are not used anywhere so both can be removed

File: YouTag/YYTTagCell.swift [hunk 23]
```diff
@@ -98,5 +106,4 @@ class YYTTagCell: UICollectionViewCell {
 			textField.isHidden = true
 		}
 	}
-
```
Comment (youstanzr) [inline YouTag/YYTTagCell.swift]: I try to have an empty line before the class closing bracket so we can keep this line

File: YouTag/ViewController.swift [hunk 24]
```diff
@@ -79,7 +79,7 @@ class ViewController: UIViewController, FilterPickerViewDelegate, YYTTagViewDele
 		filterButton.widthAnchor.constraint(equalTo: self.view.widthAnchor, multiplier: 0.2).isActive = true
 		filterButton.heightAnchor.constraint(equalTo: self.view.widthAnchor, multiplier: 0.2).isActive = true
 
-		tagsView = YYTFilterTagView(frame: .zero, tagsList: NSMutableArray(), isDeleteEnabled: true)
+		tagsView = YYTFilterTagView(frame: .zero, tagsList: NSMutableArray(), isAddEnabled: false, isMultiSelection: false, isDeleteEnabled: true, suggestionDataSource: nil)
```
Comment (youstanzr) [inline YouTag/ViewController.swift]: Function does not match the init defined for the class + YYTFilterTagView init has to be changed to handle the new `suggestionDataSource` parameter that was introduced to `super.init`

**Review by youstanzr — CHANGES_REQUESTED** [review]
Looking good 👍
Just try to fix the few comments that I left and with the proper tab indentation for code

**Comment by youstanzr** [comment]
I almost forgot, make sure the new library is added to the attributions secotion in the Readme file

**Comment by prabogey** [comment]
> Just try to fix the few comments that I left and with the proper tab indentation for code

I don't know whats wrong with the indentation, it might be some git issue. On my Xcode I don't see this indentation issue.

File: YouTag/YYTTagView.swift (hunk 20 above)
Comment (pbokey) [reply YouTag/YYTTagView.swift:30]: If the tag lists changes, how would we set the new suggestionList? The suggestion list is not based on the tagList I just pass in a suggested data source (which is just a call to getAll(.tags/.artists)) based on if the caller is requesting a YYTTagView for an artist or not.

**Comment by youstanzr** [comment]
The issue is not improper indentation as I thought but it seems like git show space indentation shorter than tab indentation which is the case in your code. The lines you added are spaces not tab indented so it might be with your Xcode formatting rules. If you prefer space indentation as your default setting for your Xcode, you can simply do the indentation manually to replace spaces by tabs.

**Comment by youstanzr** [comment]
Any update on this pull?

**Comment by prabogey** [comment]
Just sent a few commits out.

**Comment by youstanzr** [comment]
Thank you for your contribution
